### PR TITLE
makefile package flags

### PR DESCRIPTION
**Pull request by junghans**
A couple of packages add target like `CFLAGS=foo`, CC=`bar`, which is a bit hackish. With the new `make_args()`, this can be done much cleaner.

I rewrote `libxsmm` as an example.

With this in place, half of the Makefile edits could be dropped and replaced by `make CC=..`, which is much less brittle than patching.

### CODE REVIEW

File: var/spack/repos/builtin/packages/libxsmm/package.py [hunk 1]
```diff
@@ -60,22 +60,42 @@ class Libxsmm(Package):
...
-        makefile.filter('FC = ifort',       'FC ?= ifort', **kwargs)
-        makefile.filter('FC = gfortran',    'FC ?= gfortran', **kwargs)
+        return options
 
-    def manual_install(self, prefix):
+    @run_before('build')
+    def header_check(self):
         spec = self.spec
+        if '+header-only' in spec and '@1.6.2:' not in spec:
+            raise RuntimeError(
```
Comment (junghans) [inline var/spack/repos/builtin/packages/libxsmm/package.py]: @adamjstewart, I am not sure, but would `conflict('+header-only',when='@1.6.2:')` do the same?

**Comment by adamjstewart** [comment]
Hmm, so now `build_targets` and `make_args` kind of serve the same purpose, and can be used interchangeably, although I agree that they have different semantic meaning. @alalazo what are your thoughts?

**Comment by junghans** [comment]
Yes, but renaming `build_targets` everywhere, seems a bit too much ;-)

**Comment by adamjstewart** [comment]
Oh, it's not hard to rename `build_targets` everywhere. I've made changes to way more files at the same time. It's just a matter of whether we want 2 properties with similar purposes.

**Comment by adamjstewart** [comment]
Also, it should probably be called `build_args`, not `make_args`. `make` is used for both the build and install phases, and it looks like you only want to use these args during `build`.

**Comment by adamjstewart** [comment]
Another alternative to this approach is to encourage people to override `build()` instead of `build_targets` or `make_args`. I don't see anything wrong with:
```python
def build(self, spec, prefix):
    args = []

    if self.compiler.name == 'intel'
        args.append('COMPILER=intel')

    make(*args)
```

**Comment by junghans** [comment]
Wouldn't that override `build_targets`?

**Comment by adamjstewart** [comment]
Yes. `build_targets` was originally intended for projects that use something other than `all` to build, while `install_targets` was intended for projects that use something other than `install` as the target to run. You can hack `build_targets` to do whatever you want, but maybe it would make more sense to override the entire `build()` phase.

**Comment by junghans** [comment]
Ok, I guess we should:

- Document that `build_targets` and `install_targets` can be used to override `make` variables
- In code review, prefer that method over `makefile.filter()` and friends

**Comment by becker33** [comment]
If we expect to have non-target arguments to `make`, we should probably rename `build_targets` to `build_args`, and similarly for `install`.

**Comment by junghans** [comment]
@becker33 good point!

**Pull request closed without merging**